lib/delta-table/src/Database/Table/SQL/Column.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE FlexibleInstances #-}
+
+{- |
+Copyright: © 2024 Cardano Foundation
+License: Apache-2.0
+
+SQL column types.
+-}
+module Database.Table.SQL.Column
+    (
+    -- * SQL column types
+      SqlType
+    , escapeSqlType
+
+    -- * Haskell types to SQL column types
+    , IsColumn (..)
+
+    , Primary (..)
+    ) where
+
+import Prelude
+
+import Data.ByteString
+    ( ByteString
+    )
+import Data.Text
+    ( Text
+    )
+
+import qualified Database.SQLite.Simple.FromField as Sqlite
+import qualified Database.SQLite.Simple.ToField as Sqlite
+
+{-------------------------------------------------------------------------------
+    Types for database columns
+-------------------------------------------------------------------------------}
+-- | SQL column types, including constraints.
+-- Example values:
+--
+-- > INTEGER  PRIMARY KEY NOT NULL
+-- > TEXT  NOT NULL
+newtype SqlType = SqlType Text
+    deriving (Eq,Ord,Show)
+
+escapeSqlType :: SqlType -> Text
+escapeSqlType (SqlType x) = x
+
+-- | Class that maps a Haskell type to a column type of an SQL database.
+-- Includes marshalling (via the `sqlite-simple` package).
+class (Sqlite.ToField a, Sqlite.FromField a) => IsColumn a where
+    getSqlType :: proxy a -> SqlType
+
+instance IsColumn Int where
+    getSqlType _ = SqlType "INTEGER NOT NULL"
+
+instance IsColumn (Maybe Int) where
+    getSqlType _ = SqlType "INTEGER"
+
+instance IsColumn Text where
+    getSqlType _ = SqlType "TEXT NOT NULL"
+
+instance IsColumn (Maybe Text) where
+    getSqlType _ = SqlType "TEXT"
+
+instance IsColumn String where
+    getSqlType _ = SqlType "TEXT NOT NULL"
+
+instance IsColumn (Maybe String) where
+    getSqlType _ = SqlType "TEXT"
+
+instance IsColumn ByteString where
+    getSqlType _ = SqlType "BLOB NOT NULL"
+
+instance IsColumn (Maybe ByteString) where
+    getSqlType _ = SqlType "BLOB"
+
+{-------------------------------------------------------------------------------
+    Specific types
+-------------------------------------------------------------------------------}
+
+-- | Primary key.
+newtype Primary = Primary { getPrimary :: Int }
+    deriving (Eq,Ord,Show)
```

Comment:
I guess we'll likely need to support both non-int and multi-column primary keys though? https://github.com/cardano-foundation/cardano-wallet/blob/7cf4008040fc3f376fb2071d318968489b508017/lib/wallet/src/Cardano/Pool/DB/Sqlite/TH.hs#L135

... which would make it more like
```haskell
type TablePerson =
    Table "person"
        :. Col "name" Text
        :. Col "birthyear" Int
        :. PrimaryKey ["name", "birthyear"]
```
? (modulo exact syntax of `:. PrimaryKey` )

Also foreign keys...

Reply:
> I guess we'll likely need to support both non-int and multi-column primary keys though?

After some thinking, I have come to the conclusion that I don't want to support primary keys — the simplest for semantics for `Table` is the semantics where the rows are not required to satisfy additional uniqueness constraints. 🤓

These constraints provide only marginal value for us anyway — when writing code that writes to the table, we have to ensure uniqueness as an invariant, and the only thing that the SQLite uniques constraint gives us is a dynamic error message when this invariant is violated — meh.

Foreign keys are relevant for the interaction of multiple tables — which is also out of scope for `Table` at the moment.

The `Primary` type was a small concession to the possibility of having uniqueness constraint in SQLite — but I think that we are better off removing it.